proptest-derive/src/ast.rs

```diff
@@ -278,7 +279,7 @@ impl AddAssign<syn::Type> for Params {
 
 impl ToTokens for Params {
     fn to_tokens(&self, tokens: &mut TokenStream) {
-        Tuple2(self.0.as_slice()).to_tokens(tokens)
+        NestedTuple(self.0.as_slice()).to_tokens(tokens)
```

Comment:
Why change the expansion here?

(Seems like this would change the expansion of the `Parameters` associated type for cases like `enum`s as well which is a breaking change.)

Reply:
Because of the bound on `Default`on `Parameters`. `Default` is only implemented for tuples up to 12 elements currently.

How is it a breaking change for enums?

Reply:
(Thanks for reviewing this so quickly btw! :))

Reply:
> Because of the bound on `Default`on `Parameters`. `Default` is only implemented for tuples up to 12 elements currently.

Oh... that... I see.

In hindsight it might have been a mistake to auto-generate the `type Parameters` and pass values of it around without the user having asked for it specifically. It adds enormous complexity to the deriving logic and has lead to bugs. Changing the expansion for > 12 also makes it hard to have an intuition about the type `Parameters` is at all.

I would probably be best to just default to `type Parameters = ();` and then change it if the user explicitly requests so, possibly with a `#[proptest(with = "$param_expr")]` (https://github.com/AltSysrq/proptest/pull/104#discussion_r230581497). This would also do away with `no_params`.

For now we should probably keep your change as-is and then consider a more holistic breaking change with a redesign of `type Parameters` in `proptest_derive`.

cc @sunshowers and @AltSysrq if you have opinions :)

> How is it a breaking change for enums?

I re-read the code and no, it cannot be a breaking change. (My bad!) If you would generate too many parameters for an `enum` you'd run into exactly the same error as for tuples (size > 12).

> (Thanks for reviewing this so quickly btw! :))

❤️

Reply:
Yeah, I would strongly prefer the default to be `no_params` as well.

Reply:
> For now we should probably keep your change as-is and then consider a more holistic breaking change with a redesign of `type Parameters` in `proptest_derive`.

Thoughts on doing this? I'm trying to publish https://github.com/alessandrod/nuts to crates.io and thought I could depend on my own git fork but apparently crates.io doesn't allow that.

Reply:
BTW, It is a dev dependency, you could just comment it out when you publish. approximately Crater is the only user of the dev dependencies section of a published crate.